### DISCUSSION TITLE
How to use dependencies in a extension.

### DISCUSSION THREAD

**Discussion by Ncturn** [directus/directus#4279] [Archive]
Hi, im creating a module extension where i want to create a form using vue-formulate, but i dont know how to add the dependencie.

this is my rollup.config:

```js
import { terser } from 'rollup-plugin-terser';
import resolve from 'rollup-plugin-node-resolve';
import commonjs from 'rollup-plugin-commonjs';
import vue from 'rollup-plugin-vue';
import formulate from 'vue-formulate'

export default {
	input: 'src/index.js',
	output: {
		format: 'es',
		file: 'extensions/modules/microsite/index.js',
	},
	plugins: [terser(), resolve(), commonjs(), vue(), formulate],
};
```

this is my index.js:

```js
import  ModuleComponent from './microsite.vue';

export default {
	id: 'micrositio',
	name: 'micrositio',
	icon: 'box',
	routes: [
		{
			path: '/',
			component: ModuleComponent,
		},
	],
};
```

this is my component:

```vue
<template>
		<private-view title="My Custom Module">
			<formulate name="registration">
				<formulate-element
					name="email"
					type="email"
				/>
				...more formulate-elements
			</formulate>
		</private-view>
</template>

<script>
export default {
	name: 'microsite'
};
</script>
```

thanks for your time, greetings.

**Comment by rijkvanzanten** [comment]
You can `npm install` whatever you want in your extension, and use regular JS `import` in your component or js files 👍🏻

**Reply by Ncturn** [reply]
Thanks for your answer, i have to add the depencie to vue  like:
import Vue from 'vue'
import VueFormulate from '@braid/vue-formulate'

Vue.use(VueFormulate)

where can i do that?, thanks for your time.

**Reply by cpt-n3mo** [reply]
Good day!
did you find a solution for this.. i have the same problem.. doing this in index.js or in the component will make the module not load in directus.
kind regards

**Reply by nickrum** [reply]
This is not currently possible due to the way how Vue 3 works. Which dependency are you trying to use?

**Reply by cpt-n3mo** [reply]
Good day! Thx for your quick response !

That's a bummer i was trying to use vue-sse plugin. as it was available, but then i will do it myself without the plugin.
i did find that one normally should be able to do this from within the component via `this.$root.constructor.use()` but `this.$root.constructor` is undefined.. so i suspect this is not passed on to the custom module ctx.

anyways thx for the quick response.. <3 you guys are awesome fast :) 

kind regards..

**Reply by nickrum** [reply]
We might make it possible to install vue plugins in the future, but with Vue 3 and the Composition API it is usually not necessary.
I think doing `this.$root.constructor.use()` is no longer possible in Vue 3.

**Reply by cpt-n3mo** [reply]
I see, 
well then ill have to take a dive into the Composition API for abit, as i have not played with it yet, now that nuxt 3 is released its prob a good idea anyways (I mostly use nuxt). for what i have seen its quite useful, are you suggesting that plugins in general will move more to the composition api..? i guess this is more a question for vue3 then for directus...

For now i have worked around it.. so time for some reading into composition api.

kind regards.

**Reply by nickrum** [reply]
The Directus App is written using the Composition API and it works very well, especially for a large project like this. Hard to say how the Vue ecosystem will evolve, but for those plugins I have worked with, it feels like the Composition API is _the_ way to use them while supporting the Options API feels more like an afterthought for legacy purposes.

**Reply by cpt-n3mo** [reply]
Indeed, i just migrated my module, and this is quite useful indeed, basically its plane JS or TS now with some helper functions, so it keeps things nice and organized,  indeed very well suited for plugins as well. no need to inject things into ctx anymore. 
i can see why ppl are all exited about it :)

kind regards.
